index.bs

```diff
@@ -1294,44 +1296,114 @@ A {{ReceiveStream}} has the following internal slots.
+   [=ReadableStream/set up with byte reading support/pullAlgorithm=] set to |pullAlgorithm| and
+   [=ReadableStream/set up with byte reading support/cancelAlgorithm=] set to |cancelAlgorithm|.
 1. [=set/Add=] |stream| to |transport|'s [=[[ReceiveStreams]]=].
 1. Return |stream|.
 
 </div>
 
-When a {{ReceiveStream}} receives a message from the remote side aborting the
-stream (for QUIC, that message is an RESET_STREAM frame), the user agent MUST
-queue a task to run the following:
-      1. Let |stream| be the {{ReceiveStream}} object for which the reset
-         message was received.
-      1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-         from.
-      1. Let |reason| be the value from the message from the remote side.
-      1. [=ReadableStream/Error=] |stream| with |reason|.
-
-The {{ReceiveStream}}'s <dfn>cancelAlgorithm</dfn> is:
-     1. Let |stream| be the {{ReceiveStream}} object.
-     1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-        from.
-     1. Let |reason| be the first argument, if it is numeric, or 0.
-     1. If |stream| has received a RESET_STREAM frame, abort these steps.
-     1. [=In parallel=], [=send STOP_SENDING=] over |stream|'s
-        [=WebTransport stream=] with |reason| as the Application Protocol Error
-         Code.
+<div algorithm>
+
+To <dfn for="ReceiveStream">pull bytes</dfn> from a {{ReceiveStream}} |stream|, run these steps.
+
+1. Let |transport| be |stream|'s [=ReceiveStream/[[Transport]]=].
+1. Let |internalStream| be |stream|'s [=ReceiveStream/[[InternalStream]]=].
+1. Let |promise| be a new promise.
+1. Let |buffer|, |offset|, and |maxBytes| be null.
+1. If |stream|'s [=ReadableStream/current BYOB request view=] for |stream| is not null:
+  1. Set |offset| to |stream|'s [=ReadableStream/current BYOB request view=].\[[ByteOffset]].
+  1. Set |maxBytes| to |stream|'s [=ReadableStream/current BYOB request view=]'s
+     [=BufferSource/byte length=].
+  1. Set |buffer| to |stream|'s [=ReadableStream/current BYOB request view=]'s
+     [=BufferSource/underlying buffer=].
+1. Otherwise:
+  1. Set |offset| to 0.
+  1. Set |maxBytes| to an [=implementation-defined=] size.
+  1. Set |buffer| be a [=new=] {{ArrayBuffer}} with |maxBytes| size. If allocating the
+     {{Uint8Array}} fails, return a promise [=rejected with=] a {{RangeError}}.
+1. Return |promise| and run the remaining steps [=in parallel=].
+1. [=ArrayBuffer/Write=] the bytes that area [=stream/receive|read=] from |internalStream| into
+   |buffer| with offset |offset|, up to |maxBytes| bytes. Wait until either at least one byte is
+   read or FIN is received. Let |read| be the number of read bytes, and let |hasReceivedFIN| be
+   whether FIN was accompanied.
+   Note: The user-agent MAY have a buffer to improve the transfer performance. Such a buffer
+   SHOULD have a fixed size, to carry the backpressure information to the server.
```

Comment:
Can we say "fixed upper limit" instead, to allow for implementations that share a shared memory segment for several purposes, etc.

Reply:
Done.